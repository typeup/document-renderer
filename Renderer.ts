import * as dom from "@typeup/dom"
export type renderFunction = (renderer: Renderer, node: dom.Node) => Promise<string>

export class Renderer {
	private constructor(private variables: { [name: string]: string } = {}) {
	}
	getVariable(name: string): string {
		return this.variables[name]
	}
	setVariable(name: string, value: string) {
		this.variables[name] = value
	}
	async render(node: dom.Node | dom.Node[]): Promise<string> {
		let result: string
		if (Array.isArray(node))
			result = (await Promise.all(node.map(n => this.render(n)))).reduce<string[]>((r, e) => Array.isArray(e) ? [...r, ...e] : [...r, e], []).join("")
		else {
			const render = renderers[node.class]
			result = render ? await render(this, node) : ""
		}
		return result
	}
	scope(): Renderer {
		return new Renderer({ ...this.variables })
	}
	static create(): Renderer {
		return new Renderer({})
	}
}
const renderers: { [className: string]: renderFunction } = {}
export function addRenderer(className: string, render: renderFunction) {
	renderers[className] = render
}
