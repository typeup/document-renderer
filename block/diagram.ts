import yuml2svg from "yuml2svg"
import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.block.Diagram): Promise<string> {
	let value = me.value.trim()
	if (value.startsWith("//"))
		value = await yuml2svg(value)
	return `<figure>
		${ value }
		<figcaption>${ await renderer.render(me.content) }</figcaption>
	</figure>
	`
}
addRenderer("Block.Diagram", render)
