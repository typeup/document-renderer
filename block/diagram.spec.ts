import * as dom from "@typeup/dom"
import { Renderer } from "../Renderer"
import "./diagram"

jest.useFakeTimers()

const renderer = Renderer.create()
describe("Block.Diagram", () => {
	it("render", async () => {
		const node = new dom.block.Diagram(`// {type:sequence}\n[:Computer]sendUnsentEmail>[:Server]\n[:Computer]newEmail>[:Server]\n[:Server]reponse.>[:Computer]\n[:Computer]downloadEmail>[:Server]\n[:Computer]deleteOldEmail>[:Server]\n`, [new dom.inline.Text("Alice to Bob.")])
		expect(await renderer.render(node)).toMatchSnapshot()
	})
})
