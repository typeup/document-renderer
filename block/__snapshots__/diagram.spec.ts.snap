// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block.Diagram render 1`] = `
"<figure>
		<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"275.5\\" height=\\"326\\" viewBox=\\"0 0 275.5 326\\"><defs><marker id=\\"arrow-filled\\" refX=\\"6\\" refY=\\"3\\" markerWidth=\\"6\\" markerHeight=\\"6\\" orient=\\"auto\\"><path d=\\"M0,0 6,3 0,6z\\" stroke=\\"none\\" stroke-width=\\"1\\" fill=\\"#000\\"></path></marker><marker id=\\"arrow-open\\" refX=\\"6\\" refY=\\"3\\" markerWidth=\\"6\\" markerHeight=\\"6\\" orient=\\"auto\\"><path d=\\"M0,0 6,3 0,6\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"#000\\"></path></marker></defs><rect width=\\"96.5\\" height=\\"38\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" x=\\"10\\" y=\\"20\\"></rect><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"58.25\\" y=\\"39\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">:Computer</text></g><rect width=\\"96.5\\" height=\\"38\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" x=\\"10\\" y=\\"268\\"></rect><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"58.25\\" y=\\"287\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">:Computer</text></g><path d=\\"M58.25,58 v210\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\"></path><rect width=\\"79.5\\" height=\\"38\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" x=\\"166\\" y=\\"20\\"></rect><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"205.75\\" y=\\"39\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">:Server</text></g><rect width=\\"79.5\\" height=\\"38\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" x=\\"166\\" y=\\"268\\"></rect><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"205.75\\" y=\\"287\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">:Server</text></g><path d=\\"M205.75,58 v210\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\"></path><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"132\\" y=\\"83\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">sendUnsentEmail</text></g><path d=\\"M58.25,96 h147.5\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" marker-end=\\"url(#arrow-filled)\\"></path><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"132\\" y=\\"121\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">newEmail</text></g><path d=\\"M58.25,134 h147.5\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" marker-end=\\"url(#arrow-filled)\\"></path><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"132\\" y=\\"159\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">reponse</text></g><path d=\\"M205.75,172 h-147.5\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" stroke-dasharray=\\"7,4\\" marker-end=\\"url(#arrow-filled)\\"></path><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"132\\" y=\\"197\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">downloadEmail</text></g><path d=\\"M58.25,210 h147.5\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" marker-end=\\"url(#arrow-filled)\\"></path><g><text fill=\\"#000\\" font-family=\\"Helvetica,sans-Serif\\" x=\\"132\\" y=\\"235\\" text-anchor=\\"middle\\" alignment-baseline=\\"central\\">deleteOldEmail</text></g><path d=\\"M58.25,248 h147.5\\" stroke=\\"#000\\" stroke-width=\\"1\\" fill=\\"none\\" marker-end=\\"url(#arrow-filled)\\"></path></svg>
		<figcaption></figcaption>
	</figure>
	"
`;
